source/adios2/helper/adiosXML.cpp

```diff
@@ -11,473 +11,162 @@
+                throw std::runtime_error(
+                    "No \"value\" attribute found on <parameter> element.");
             }
-            else
-            {
-                throw std::invalid_argument(
-                    "ERROR: quote must be \" or ' in XML config tag " + tag +
-                    ", in call to ADIOS constructor");
-            }
-
-            attributes.emplace(key, value);
-            currentPosition = quotePosition + value.size() + 1;
-        }
-        return attributes;
-    };
-
-    // BODY of function starts here
-    Params attributes;
-    // eliminate < >
-    std::string openingTag = tagHeader.substr(1, tagHeader.size() - 2);
-
-    if (tagHeader.back() == '/') // last char is / --> "XML empty tag"
-    {
-        // attributes = lf_GetAttributes(openingTag);
-        // throw exception here, ADIOS2 doesn't allow XML empty tags
-    }
-    else if (tagHeader[0] == '/') // first char is / ---> closing tag
-    {
-        attributes = lf_GetAttributes(openingTag);
-        if (attributes.size() > 0)
-        {
-            throw std::invalid_argument(
-                "ERROR: closing tag " + tagHeader +
-                " can't have attributes, in call to ADIOS constructor\n");
-        }
-    }
-    else // opening tag
-    {
-        attributes = lf_GetAttributes(openingTag);
-    }
-    return attributes;
-}
-
-void InitXML(const std::string configXML, const MPI_Comm mpiComm,
-             const bool debugMode,
-             std::vector<std::shared_ptr<Transform>> &transforms,
-             std::map<std::string, IO> &ios)
-{
-    // independent IO
-    std::string fileContents(FileToString(configXML));
-    if (fileContents.empty())
-    {
-        // issue a warning?
-        return;
-    }
-
-    RemoveCommentsXML(fileContents);
-
-    // adios-config
-    std::string::size_type position(0);
-    const TagXML adiosConfigXML(
-        GetTagXML("adios-config", fileContents, position));
-
-    // process transforms, not yet implemented
-
-    while (position != std::string::npos)
-    {
-        const TagXML transformXML(
-            GetTagXML("transform ", adiosConfigXML.Elements, position));
-
-        if (transformXML.Header.empty())
-        {
-            break;
+            continue;
         }
-        // InitTransform(transformTag, debugMode, transforms);
-    }
-
-    position = 0;
-    // process IOs
-    while (position != std::string::npos)
-    {
-        // io
-        const TagXML ioXML(GetTagXML("io ", adiosConfigXML.Elements, position));
 
-        if (ioXML.Header.empty()) // no more groups to find
-        {
-            break;
-        }
-        InitIOXML(ioXML, mpiComm, debugMode, transforms, ios);
+        params.emplace(attrKey.value(), attrValue.value());
     }
+    return params;
 }
 
-void InitIOXML(const TagXML &ioXML, const MPI_Comm mpiComm,
+void InitIOXML(pugi::xml_node ioNode, const MPI_Comm mpiComm,
```

Comment:
Const correctness matters all over the code. Provides intent and compile time checks. Has nothing to do with scope.